### PR TITLE
Edify Webhook

### PR DESCRIPTION
**Pull request by nateh777**
##### What does the change add or fix?

Create Edify webhook for open and link tracking for communications.

---------

##### What does the change add or fix in a succinct statement that will be read by clients?

- Create Edify webhook for open and link tracking for communications.

---------

##### Who reported, requested, or paid for the change?

Edify

---------

##### Does this update or add options to the block UI?

image

---------

##### What files does it affect?

Webhooks/Edify.ashx

---------

##### Is it a breaking change for other versions/clients?

No

### CODE REVIEW

File: Webhooks/Edify.ashx [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+
+                var interactionService = new InteractionService( rockContext );
+                DateTime timeStamp = RockDateTime.ConvertLocalDateTimeToRockDateTime( new DateTime( 1970, 1, 1, 0, 0, 0, DateTimeKind.Utc ).AddSeconds( edifyEvent.Timestamp.Value ).ToLocalTime() );
+
+                switch ( edifyEvent.Event )
+                {
+                    case "processed":
+                        // Do nothing.
+                        break;
+                    case "dropped":
```
Comment (nateh777) [inline Webhooks/Edify.ashx:271]:
```suggestion
```

File: Webhooks/Edify.ashx [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+                        else
+                        {
+                            Rock.Communication.Email.ProcessBounce(
+                                edifyEvent.Payload.Message.To,
+                                Rock.Communication.BounceType.SoftBounce,
+                                communicationRecipient.StatusNote,
+                                RockDateTime.ConvertLocalDateTimeToRockDateTime( new DateTime( 1970, 1, 1, 0, 0, 0, DateTimeKind.Utc ).AddSeconds( edifyEvent.Timestamp.Value ).ToLocalTime() ) );
+                        }
+                        break;
+                    case "delivered":
```
Comment (nateh777) [inline Webhooks/Edify.ashx:302]:
```suggestion
```

File: Webhooks/Edify.ashx [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+                    case "MessageSent":
+                        communicationRecipient.Status = CommunicationRecipientStatus.Delivered;
+                        communicationRecipient.StatusNote = string.Format( "Confirmed delivered by Edify at {0}", timeStamp.ToString() );
+                        break;
+                    case "MessageDelayed":
+                        // TODO: handle MessageDelayed.
+                        break;
+                    case "bounce":
```
Comment (nateh777) [inline Webhooks/Edify.ashx:310]:
```suggestion
```

File: Webhooks/Edify.ashx [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+                                RockDateTime.ConvertLocalDateTimeToRockDateTime( new DateTime( 1970, 1, 1, 0, 0, 0, DateTimeKind.Utc ).AddSeconds( edifyEvent.Timestamp.Value ).ToLocalTime() ) );
+                        }
+                        break;
+                    case "delivered":
+                    case "MessageSent":
+                        communicationRecipient.Status = CommunicationRecipientStatus.Delivered;
+                        communicationRecipient.StatusNote = string.Format( "Confirmed delivered by Edify at {0}", timeStamp.ToString() );
+                        break;
+                    case "MessageDelayed":
+                        // TODO: handle MessageDelayed.
```
Comment (nateh777) [inline Webhooks/Edify.ashx:308]: Does `CommunicationRecipientStatus.Pending` have anything else attached to it? Like actually sending a message.

File: Webhooks/Edify.ashx [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+                        break;
+                    case "bounce":
+                    case "MessageBounced":
+                        communicationRecipient.Status = CommunicationRecipientStatus.Failed;
+                        communicationRecipient.StatusNote = edifyEvent.Payload.Details + edifyEvent.Payload.Output;
```
Comment (nateh777) [inline Webhooks/Edify.ashx:313]: Work on this status note, make a more dynamic string here

File: Webhooks/Edify.ashx [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+                        //communicationRecipient.StatusNote = edifyEvent.Payload.Bounce.Subject;
+
+                        Rock.Communication.Email.ProcessBounce(
+                            edifyEvent.Payload.Message.To,
+                            Rock.Communication.BounceType.HardBounce,
+                            edifyEvent.Event,
+                            timeStamp );
+                        break;
+                    case "open":
```
Comment (nateh777) [inline Webhooks/Edify.ashx:322]:
```suggestion
```

File: Webhooks/Edify.ashx [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+
+        private void ProcessForWorkflow( Guid? actionGuid, Rock.Data.RockContext rockContext, EdifyEvent edifyEvent )
+        {
+            RockLogger.Log.Debug( RockLogDomains.Communications, "ProcessForWorkflow {@payload}", edifyEvent );
+
+            string status = string.Empty;
+            switch ( edifyEvent.Event )
+            {
+                //case "unsubscribe":
+                //case "delivered":
```
Comment (nateh777) [inline Webhooks/Edify.ashx:392]:
```suggestion
```

File: Webhooks/Edify.ashx [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+                case "MessageSent":
+                    status = SendEmailWithEvents.SENT_STATUS;
+                    break;
+
+                //case "click":
```
Comment (nateh777) [inline Webhooks/Edify.ashx:397]:
```suggestion
```

File: Webhooks/Edify.ashx [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+                case "MessageLinkClicked":
+                    status = SendEmailWithEvents.CLICKED_STATUS;
+                    break;
+
+                //case "open":
```
Comment (nateh777) [inline Webhooks/Edify.ashx:402]:
```suggestion
```

File: Webhooks/Edify.ashx [hunk 10]
```diff
@@ -0,0 +1,455 @@
...
+                case "MessageLoaded":
+                    status = SendEmailWithEvents.OPENED_STATUS;
+                    break;
+
+                case "failed":
+                case "dropped":
+                case "blocked":
+                case "bounce":
+                case "bounced":
```
Comment (nateh777) [inline Webhooks/Edify.ashx:411]:
```suggestion
```

File: Webhooks/Edify.ashx [hunk 11]
```diff
@@ -0,0 +1,455 @@
...
+                                message,
+                                RockDateTime.ConvertLocalDateTimeToRockDateTime( new DateTime( 1970, 1, 1, 0, 0, 0, DateTimeKind.Utc ).AddSeconds( edifyEvent.Timestamp.Value ).ToLocalTime() ) );
+                    }
+                    break;
+                case "MessageBounced":
+                    status = SendEmailWithEvents.FAILED_STATUS;
+                    string messageBounce = edifyEvent.Payload.Output.IsNotNullOrWhiteSpace() ? edifyEvent.Payload.Output : edifyEvent.Payload.Details;
+                    if ( messageBounce.IsNullOrWhiteSpace() )
+                    {
+                        messageBounce = string.Format( "{0}. Message Bounced event.", edifyEvent.Payload.Bounce.Subject );
```
Comment (nateh777) [inline Webhooks/Edify.ashx:439]: Match with string build you make further up.

File: Webhooks/Edify.ashx (hunk 4 above)
Comment (nateh777) [reply Webhooks/Edify.ashx:303]: Did a quick search and I am not comfortable with setting it to Pending based on what I saw. Though I didn't confirm that is how it sends there are at least a couple methods that look for pending communications that then call a send method. So we will just leave it for now.

**Comment by nateh777** [comment]
@guymaness2 made the modifications from our merge review and it should be ready to go now.

**Review by guymaness2 — APPROVED** [review]

**Pull request merged**